dps310/dps310.go

```diff
@@ -0,0 +1,539 @@
+	err = d.writeRegister(RESET, RESET_VAL)
+	if err != nil {
+		return
+	}
+
+	// Wait for a bit after reset
+	time.Sleep(10 * time.Millisecond)
+
+	// Wait for Sensor getting ready
+	rdy := false
+	for {
+		rdy, err = sensorRdy()
+		if err != nil {
+			return
+		}
+		if rdy {
+			break
+		}
+		time.Sleep(20 * time.Millisecond)
+	}
+	return
+}
+
+func (d *DPS310) readCalibration() (err error) {
+
+	// check func for calibration ready
+	calibRdy := func() (bool, error) {
+		v, err := d.readRegister(MEAS_CFG)
+		if err != nil {
+			return false, err
+		}
+		return v & MEAS_CFG_COEF_RDY_MSK != 0, nil
+	}
+
+	// Wait till we're ready to read calibration
+	rdy := false
+	for {
+		rdy, err = calibRdy()
+		if err != nil {
+			return
+		}
+		if rdy {
+			break
+		}
+		time.Sleep(20 * time.Millisecond)
+	}
+
+	var (
+		coeffs [18]uint8
+	)
+	i := 0
+	for addr := uint8(COEF_START); addr < COEF_END; addr++ {
+		coeffs[i], err = d.readRegister(addr)
+		if err != nil {
+			return
+		}
+		i++
+	}
+
+	// see: datasheet page 37
+	d.c0 = decVal((uint12_t(coeffs[0])<<4 |
+		(uint12_t(coeffs[1]) >> 4 & 0x0f)))
+
+	d.c1 = decVal((uint12_t(coeffs[1])&0xf)<<8 |
+		uint12_t(coeffs[2]))
+
+	d.c00 = decVal((uint20_t(coeffs[3]) << 12) |
+		(uint20_t(coeffs[4]) << 4) |
+		((uint20_t(coeffs[5]) >> 4) & 0x0F))
+
+	d.c10 = decVal(((uint20_t(coeffs[5]) & 0x0f) << 16) |
+		(uint20_t(coeffs[6]) << 8) |
+		uint20_t(coeffs[7]))
+
+	d.c01 = decVal((uint16_t(coeffs[8]) << 8) |
+		uint16_t(coeffs[9]))
+
+	d.c11 = decVal((uint16_t(coeffs[10]) << 8) |
+		uint16_t(coeffs[11]))
+
+	d.c20 = decVal((uint16_t(coeffs[12]) << 8) |
+		uint16_t(coeffs[13]))
+
+	d.c21 = decVal((uint16_t(coeffs[14]) << 8) |
+		uint16_t(coeffs[15]))
+
+	d.c30 = decVal((uint16_t(coeffs[16]) << 8) |
+		uint16_t(coeffs[17]))
+
+	return
+}
+
+type uint12_t uint32
+type uint16_t uint32
+type uint20_t uint32
+type uint24_t uint32
+
+type coefs interface {
+	uint12_t | uint16_t | uint20_t | uint24_t
+}
+
+func decVal[T coefs](v T) float32 {
+	const (
+		max12 = 1<<11 - 1
+		sub12 = 1 << 12
+		max16 = 1<<15 - 1
+		sub16 = 1 << 16
+		max20 = 1<<19 - 1
+		sub20 = 1 << 20
+		max24 = 1<<23 - 1
+		sub24 = 1 << 24
+	)
+
+	var (
+		max, sub int32
+	)
+
+	switch any(v).(type) {
+	case uint12_t:
+		max = max12
+		sub = sub12
+	case uint16_t:
+		max = max16
+		sub = sub16
+	case uint20_t:
+		max = max20
+		sub = sub20
+	case uint24_t:
+		max = max24
+		sub = sub24
+	}
+
+	val := int32(v)
+	if val > max {
+		val -= sub
+	}
+	return float32(val)
+}
+
+// Set the sample rate and oversampling averaging for pressure and temperature
+// rate How many samples per second to take
+// param os How many oversamples to average
+func (d *DPS310) ConfigureValues(rate Rate, os Precision) {
+	d.ConfigurePressure(rate, os)
+	d.ConfigureTemperature(rate, os)
+}
+
+func (d *DPS310) isOkRateOs() bool {
+	return float32(d.pressureRate)*osScalingData[d.pressureOs].measTime+
+		float32(d.tempRate)*osScalingData[d.tempOs].measTime >= 1.0
+}
```

Comment:
I think this check should not look at values stored on the struct, but instead take the values as argument. 
Otherwise you would need to restore the values to the previous one if the chosen combination is not okay and the change is not performed in the functions below